### ISSUE TITLE
Ports are always closed when running SIF built from docker via instance run

### ISSUE THREAD

**Issue by ntippman** [sylabs/singularity#2629]
**Version of Singularity**
`singularity-ce version 4.1.0-1`

**Describe the bug and steps to reproduce**
_TLDR: Ports are always closed when running SIF generated from docker image via `singularity instance start`. However, running the same SIF with `singularity run` works as expected._

I am running multiple docker containers in a compose infrastructure and now want to add support for singularity. Unfortunately, I am running into issues with the network as mapped ports are always closed. 

My setup is to generate docker images from my dockerfiles and then convert them to SIF.
```
docker build -t myimg -f ./docker/myimage.dockerfile .
singularity build myimage.sif docker-daemon://myimage:latest
```
For demonstration purposes I will use the official nginx docker container.

```
docker pull nginx
singularity build nginx.sif docker-daemon://nginx
```

When starting the image as an instance, the port seems to be closed, regardless of `portmap` usage. All commands are run as `root`.

```
$ singularity instance start --writable-tmpfs nginx.sif test1
$ singularity instance start --writable-tmpfs --net --network-args "portmap=8000:80/tcp" nginx.sif test2

$ singularity instance list
INSTANCE NAME    PID       IP            IMAGE
test1            224960                  /path/nginx.sif
test2            225000    10.22.0.18    /path/nginx.sif

# no instance reachable with curl

$ curl localhost:80
curl: (7) Failed to connect to localhost port 80: Connection refused
$ curl localhost:8000
curl: (7) Failed to connect to localhost port 8000: Connection refused
$ curl 10.22.0.18:8000
curl: (7) Failed to connect to 10.22.0.18 port 8000: Connection refused
```

I am not sure if there is a fundamental difference in network handling between `instance start` and `run`, but when I start the same SIF with `run` all ports are available as expected:

```
$ singularity run --writable-tmpfs --net --network-args "portmap=8001:80/tcp" nginx.sif

# different shell

$ curl localhost:8001
<!DOCTYPE html>
<html>
<head>
<title>Welcome to nginx!</title>
...
```

When I tried the nginx example from [this guide](https://docs.sylabs.io/guides/3.5/user-guide/running_services.html#nginx-hello-world-in-singularity) everything works as expected:

```
#### nginx.def
Bootstrap: docker
From: nginx
Includecmd: no

%startscript
   nginx

###

$ singularity build nginx_from_def.sif nginx.def
$ singularity instance start --writable-tmpfs --net --network-args "portmap=8002:80/tcp" nginx_from_def.sif test3

$ curl localhost:8002
<!DOCTYPE html>
<html>
<head>
<title>Welcome to nginx!</title>
...
```

Examining `iptables` shows no difference between all of the different variants (or I failed to spot them).

```
$ iptables-save
# Generated by iptables-save v1.8.8 (nf_tables) on Fri Feb  2 14:37:23 2024
*nat
:PREROUTING ACCEPT [0:0]
:INPUT ACCEPT [0:0]
:OUTPUT ACCEPT [0:0]
:POSTROUTING ACCEPT [0:0]
:CNI-84e3116b19e0b2beee4b3bf4 - [0:0]
:CNI-98ba8cb7764ed8b5ac79683b - [0:0]
:CNI-DN-84e3116b19e0b2beee4b3 - [0:0]
:CNI-DN-98ba8cb7764ed8b5ac796 - [0:0]
:CNI-DN-b6bbbf987fffae51e1817 - [0:0]
:CNI-DN-f8fe102ebdfdf568b17b2 - [0:0]
:CNI-HOSTPORT-DNAT - [0:0]
:CNI-HOSTPORT-MASQ - [0:0]
:CNI-HOSTPORT-SETMARK - [0:0]
:CNI-b6bbbf987fffae51e1817e40 - [0:0]
:CNI-f8fe102ebdfdf568b17b23e0 - [0:0]
-A PREROUTING -m addrtype --dst-type LOCAL -j CNI-HOSTPORT-DNAT
-A OUTPUT -m addrtype --dst-type LOCAL -j CNI-HOSTPORT-DNAT
-A POSTROUTING -m comment --comment "CNI portfwd requiring masquerade" -j CNI-HOSTPORT-MASQ
-A POSTROUTING -s 10.22.0.10/32 -m comment --comment "name: \"bridge\" id: \"222287\"" -j CNI-98ba8cb7764ed8b5ac79683b
-A POSTROUTING -s 10.22.0.18/32 -m comment --comment "name: \"bridge\" id: \"225000\"" -j CNI-84e3116b19e0b2beee4b3bf4
-A POSTROUTING -s 10.22.0.21/32 -m comment --comment "name: \"bridge\" id: \"225600\"" -j CNI-f8fe102ebdfdf568b17b23e0
-A POSTROUTING -s 10.22.0.22/32 -m comment --comment "name: \"bridge\" id: \"225743\"" -j CNI-b6bbbf987fffae51e1817e40
-A CNI-84e3116b19e0b2beee4b3bf4 -d 10.22.0.0/16 -m comment --comment "name: \"bridge\" id: \"225000\"" -j ACCEPT
-A CNI-84e3116b19e0b2beee4b3bf4 ! -d 224.0.0.0/4 -m comment --comment "name: \"bridge\" id: \"225000\"" -j MASQUERADE
-A CNI-98ba8cb7764ed8b5ac79683b -d 10.22.0.0/16 -m comment --comment "name: \"bridge\" id: \"222287\"" -j ACCEPT
-A CNI-98ba8cb7764ed8b5ac79683b ! -d 224.0.0.0/4 -m comment --comment "name: \"bridge\" id: \"222287\"" -j MASQUERADE
-A CNI-DN-84e3116b19e0b2beee4b3 -s 10.22.0.0/16 -p tcp -m tcp --dport 8000 -j CNI-HOSTPORT-SETMARK
-A CNI-DN-84e3116b19e0b2beee4b3 -s 127.0.0.1/32 -p tcp -m tcp --dport 8000 -j CNI-HOSTPORT-SETMARK
-A CNI-DN-84e3116b19e0b2beee4b3 -p tcp -m tcp --dport 8000 -j DNAT --to-destination 10.22.0.18:80
-A CNI-DN-98ba8cb7764ed8b5ac796 -s 10.22.0.0/16 -p tcp -m tcp --dport 8889 -j CNI-HOSTPORT-SETMARK
-A CNI-DN-98ba8cb7764ed8b5ac796 -s 127.0.0.1/32 -p tcp -m tcp --dport 8889 -j CNI-HOSTPORT-SETMARK
-A CNI-DN-98ba8cb7764ed8b5ac796 -p tcp -m tcp --dport 8889 -j DNAT --to-destination 10.22.0.10:80
-A CNI-DN-b6bbbf987fffae51e1817 -s 10.22.0.0/16 -p tcp -m tcp --dport 8001 -j CNI-HOSTPORT-SETMARK
-A CNI-DN-b6bbbf987fffae51e1817 -s 127.0.0.1/32 -p tcp -m tcp --dport 8001 -j CNI-HOSTPORT-SETMARK
-A CNI-DN-b6bbbf987fffae51e1817 -p tcp -m tcp --dport 8001 -j DNAT --to-destination 10.22.0.22:80
-A CNI-DN-f8fe102ebdfdf568b17b2 -s 10.22.0.0/16 -p tcp -m tcp --dport 8002 -j CNI-HOSTPORT-SETMARK
-A CNI-DN-f8fe102ebdfdf568b17b2 -s 127.0.0.1/32 -p tcp -m tcp --dport 8002 -j CNI-HOSTPORT-SETMARK
-A CNI-DN-f8fe102ebdfdf568b17b2 -p tcp -m tcp --dport 8002 -j DNAT --to-destination 10.22.0.21:80
-A CNI-HOSTPORT-DNAT -p tcp -m comment --comment "dnat name: \"bridge\" id: \"222287\"" -m multiport --dports 8889 -j CNI-DN-98ba8cb7764ed8b5ac796
-A CNI-HOSTPORT-DNAT -p tcp -m comment --comment "dnat name: \"bridge\" id: \"225000\"" -m multiport --dports 8000 -j CNI-DN-84e3116b19e0b2beee4b3
-A CNI-HOSTPORT-DNAT -p tcp -m comment --comment "dnat name: \"bridge\" id: \"225600\"" -m multiport --dports 8002 -j CNI-DN-f8fe102ebdfdf568b17b2
-A CNI-HOSTPORT-DNAT -p tcp -m comment --comment "dnat name: \"bridge\" id: \"225743\"" -m multiport --dports 8001 -j CNI-DN-b6bbbf987fffae51e1817
-A CNI-HOSTPORT-MASQ -m mark --mark 0x2000/0x2000 -j MASQUERADE
-A CNI-HOSTPORT-SETMARK -m comment --comment "CNI portfwd masquerade mark" -j MARK --set-xmark 0x2000/0x2000
-A CNI-b6bbbf987fffae51e1817e40 -d 10.22.0.0/16 -m comment --comment "name: \"bridge\" id: \"225743\"" -j ACCEPT
-A CNI-b6bbbf987fffae51e1817e40 ! -d 224.0.0.0/4 -m comment --comment "name: \"bridge\" id: \"225743\"" -j MASQUERADE
-A CNI-f8fe102ebdfdf568b17b23e0 -d 10.22.0.0/16 -m comment --comment "name: \"bridge\" id: \"225600\"" -j ACCEPT
-A CNI-f8fe102ebdfdf568b17b23e0 ! -d 224.0.0.0/4 -m comment --comment "name: \"bridge\" id: \"225600\"" -j MASQUERADE
COMMIT
```

I'm at my wit's end and no longer sure if this is a problem on my end or Singularity's side.

**Expected behavior**
Ports should be open regardless of startup via `instance start` or `run`.

**OS / Linux Distribution**
`AlmaLinux 9.3`

**Installation Method**
`dnf install -y https://github.com/sylabs/singularity/releases/download/v4.1.0/singularity-ce-4.1.0-1.el9.x86_64.rpm`

**Comment by dtrudg** [comment]
When you `run` a container then the `%runscript` used to build the container is used.

When you `instance start` a container then the `%startscript` used to build the container is used.

If you directly convert a Docker container to a Singularity container, only a `%runscript` is populated (from the Docker container CMD / ENTRYPOINT). There is no `%startscript`, so `instance start` doesn't start anything.

You will need to build with a definition file that adds a `%startscript`. The example from the docs works, as it defines a `%startscript`.

**Comment by ntippman** [comment]
Well... that's good to know. I never questioned whether the application inside the container was running since I expected the same behaviour as with docker - that the container stops once PID 1 exits.

Thank you very much!
